aten/src/ATen/native/cpu/UnaryOpsKernel.cpp

```diff
@@ -348,10 +348,13 @@ static void trigamma_kernel(TensorIterator& iter) {
 }
 
 static void polygamma_kernel(TensorIterator& iter, int64_t n) {
-  switch (n) {
-    case 0: digamma_kernel(iter); break;
-    case 1: trigamma_kernel(iter); break;
```

Comment:
Why don't you use the specialized kernel for trigamma anymore? Can it be deleted if it is not used anymore?

Reply:
Based on your comment above, this is not strictly needed. So I think what we should do is:
- if trigamma is faster than the general kernel, use it.
- If not, delete the code

Reply:
From my knowledge, digamma is slightly differently implemented with polygamma with n=0 thus `digamma_kernel `may run faster than `polygamma_kernel(0, x)`

But `trigamma_kernel` has almost same with polygamma, with n=1 constant. thus i thought it was redundant but I didn't check performance comparison at all.

Reply:
Sounds good.
It would be nice if you can run a very quick timing to confirm that (don't forget torch.cuda.synchronize() when running GPU timings).
And then we can remove the unused code 🎉

Reply:
I ran this code; 
```python
x = torch.randn(100,100)
with torch.autograd.profiler.profile() as prof:
    for i in range(10000):
        t = torch.polygamma(1, x)


print(prof.key_averages().table())


torch.cuda.synchronize()
x = torch.randn(100,100).cuda()
with torch.autograd.profiler.profile(use_cuda=True) as prof:
    for i in range(10000):
        t = torch.polygamma(1, x)


print(prof.key_averages().table())
```

### Trigamma used

#### CPU ver
>>> print(prof.key_averages().table())
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
polygamma      100.00%          3.781s           100.00%          3.781s           378.112us        10000
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 3.781s



#### cuda ver

>>> print(prof.key_averages().table())
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name           Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
polygamma      100.00%          323.681ms        100.00%          323.681ms        32.368us         100.00%          350.363ms        35.036us         10000
-------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 323.681ms
CUDA time total: 350.363ms

### Polygamma used

I run only 1000 iters. (trigamma run 10,000 iters)
My polygamma implementations took about 20 times slower than trigamma (very far from what I've expected to be equal).

#### cpu ver
>>> print(prof.key_averages().table())
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     Number of Calls
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
aten::polygamma      99.95%           8.074s           199.88%          16.145s          8.073ms          2000
aten::empty          0.03%            2.104ms          0.03%            2.104ms          2.104us          1000
aten::resize_        0.02%            1.557ms          0.02%            1.557ms          1.557us          1000
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 8.078s

#### cuda ver
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                 Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
aten::polygamma      88.35%           198.364ms        180.65%          405.621ms        202.811us        97.79%           467.375ms        233.687us        2000
aten::empty          3.23%            7.242ms          3.23%            7.242ms          7.242us          1.01%            4.831ms          4.831us          1000
aten::resize_        8.43%            18.924ms         8.43%            18.924ms         18.924us         1.20%            5.742ms          5.742us          1000
-------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 224.530ms
CUDA time total: 477.948ms

Reply:
So, It is better to save trigamma as you noted!

Reply:
Ok nice.
In that case, I think you can leave the special case for trigamma, just like the one for digamma. And use your function only for n >= 2?

Reply:
Yes. I'll patch soon.